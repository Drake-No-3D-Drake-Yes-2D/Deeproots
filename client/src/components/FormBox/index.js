import React, { Component } from 'react'
import axios from 'axios'
import './index.css'

const GOOGLE_FORM_MESSAGE_ID = 'entry.2606285'
const GOOGLE_FORM_ATTENDING_ID = 'entry.1498135098'
const CORS_PROXY = 'https://cors-anywhere.herokuapp.com/'
const GOOGLE_FORM_ACTION = 'https://docs.google.com/forms/u/5/d/e/1FAIpQLSfBYrkGVJrlkgxuGVzzBTnGOCJpVEO4b8vhDNXEr4SAk8Ei7w/formResponse'

class FormBox extends Component {
  constructor(props){
    super(props)
    this.state = { 
      message: '', 
      attending: '', 
      showForm: false,
      sendingMessage: false,
      messageSent: false,
      messageError: false
    }
  }

  handleChange = (event) => {
    this.setState({[event.target.name]: event.target.value})
  }

  handleSubmit = (event) => {
    event.preventDefault()
    this.setState({
      sendingMessage: true
    })
    this.sendMessage()
  }

  handleFormToggle = () => {
    this.setState( 
      (prevState) => {
        const { showForm } = prevState
        return {
          showForm: !showForm
        } 
      }
    )
  }

  handleReturnButton = () => {
    this.setState(
      {
        showForm: false,
        messageSent: false,
        messageError: false
      }
    )
  }

  sendMessage = () => {
    const formData = new FormData()
    formData.append(GOOGLE_FORM_MESSAGE_ID, this.state.message)
    formData.append(GOOGLE_FORM_ATTENDING_ID, this.state.attending)

    axios.post(CORS_PROXY + GOOGLE_FORM_ACTION,formData)
      .then(() => {
        this.setState({ 
          messageSent: true,
          sendingMessage: false,
          message: '',
          attending: ''
        })
      }).catch(() => {
        this.setState({ 
          messageError: true,
          sendingMessage: false
        })
      })
  }

  returnButton = () => <button id='return-button' className='btn btn-default btn-xs' onClick={this.handleReturnButton}>Return</button>

  render() {
    if(this.state.sendingMessage){
      return(
        <div>Sending...</div>
      )
    }

    if(this.state.messageSent){
      return(
        <React.Fragment>
          <div className='success-message'>You are successfully Registered</div>
          {this.returnButton()}
        </React.Fragment>
      )
    }

    if(this.state.messageError){
      return(
        <React.Fragment>
          <div className='error-message'>Sorry, you were not registered. Please Try again</div>
          {this.returnButton()}
        </React.Fragment>
      )
    }

    if(this.state.showForm === false){
      return(
        <button id='contact-button' className='btn btn-sm' onClick={this.handleFormToggle}>Sign Up</button>
      )
    }

    return(
      <React.Fragment>
        <div className='form-container'>
          <form onSubmit={this.handleSubmit}>
            <div className='form-group row'>
              <label htmlFor='attending' className='col-sm-2 col-form-label'>
                What are the names of people attending?:
              </label>
              <div className='col-sm-8'>
                <textarea 
                  type='attending' 
                  name='attending'
                  id='attending'
                  className='form-control'
                  value={this.state.attending} 
                  onChange={this.handleChange}
                  required
                  rows='2'
                />
              </div>
            </div>
            <div className='form-group row'>
              <label htmlFor='message' className='col-sm-2 col-form-label'>
                Comments and/or questions:
              </label>
              <div className='col-sm-8'>
                <textarea 
                  id='message'
                  name='message'
                  className='form-control'
                  value={this.state.message} 
                  onChange={this.handleChange} 
                  rows='4'
                />
              </div>
            </div>
            <div>
              <button type='button' id='cancel-button' className='btn btn-default btn-sm btn-action' onClick={this.handleFormToggle}>Cancel</button>
              <button type='submit' className='btn btn-sm btn-default btn-action'>Submit</button>
            </div>
          </form>
        </div>
      </React.Fragment>
    )
  }
}

export default FormBox