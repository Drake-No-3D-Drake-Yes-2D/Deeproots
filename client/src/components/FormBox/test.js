import React from 'react'
import { shallow } from 'enzyme'
import moxios from 'moxios'
import FormBox from '.'

describe('form Box', () => {
  let formBox

  beforeEach(() => {
    formBox = shallow(<FormBox />)
    moxios.install()
  })

  afterEach(() => {
    moxios.uninstall()
  })

  it('Renders correctly', () => {
    expect(formBox).toMatchSnapshot()
  })

  it('renders with the form hidden by default', () => {
    expect(formBox.find('form').length).toBe(0)
  })

  it('renders with the form button by default', () => {
    expect(formBox.find('#form-button').length).toBe(1)
  })

  it('shows form when the form button is clicked ', () => {
    formBox.find('#form-button').simulate('click')

    expect(formBox.find('form').length).toBe(1)
  })

  it('hides form when the cancel button is clicked', () => {
    formBox.setState({showForm: true})
    formBox.find('#cancel-button').simulate('click')

    expect(formBox.find('form').length).toBe(0)
  })

  it('shows form button when the cancel button is clicked', () => {
    formBox.setState({showForm: true})
    formBox.find('#cancel-button').simulate('click')

    expect(formBox.find('#form-button').length).toBe(1)
  })

  it('shows loading message after submitting form', () => {
    formBox.setState({showForm: true})
    formBox.find('form').simulate('submit', {
      preventDefault: jest.fn()
    })

    expect(formBox.text()).toBe('Sending...')
  })

  it('shows success message and form button if the message is sent successfully', () => {
    formBox.setState({messageSent: true})

    expect(formBox.find('.success-message').length).toBe(1)
    expect(formBox.find('#return-button').length).toBe(1)
  })

  it('displays form button when the return button is clicked', () => {
    formBox.setState({messageSent: true})
    formBox.find('#return-button').simulate('click')

    expect(formBox.find('#form-button').length).toBe(1)
  })

  it('displays error message and return button when message is not sent', () => {
    formBox.setState({messageError: true})

    expect(formBox.find('.error-message').length).toBe(1)
    expect(formBox.find('#return-button').length).toBe(1)
  })
})