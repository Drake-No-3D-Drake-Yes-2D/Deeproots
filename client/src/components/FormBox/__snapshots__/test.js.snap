// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Box Renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FormBox />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "Form",
      "className": "btn btn-sm",
      "id": "contact-button",
      "onClick": [Function],
    },
    "ref": null,
    "rendered": "Form",
    "type": "button",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Form",
        "className": "btn btn-sm",
        "id": "contact-button",
        "onClick": [Function],
      },
      "ref": null,
      "rendered": "Form",
      "type": "button",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
